contracts/router/LinkedPool.sol

```diff
@@ -0,0 +1,377 @@
+        }
+    }
+
+    // ══════════════════════════════════════════════ INTERNAL VIEWS ═══════════════════════════════════════════════════
+
+    /// @dev Returns the amount of tokens that will be received from a single swap.
+    function _getPoolQuote(
+        address poolModule,
+        address pool,
+        uint256 nodeIndexFrom,
+        uint256 nodeIndexTo,
+        uint256 amountIn,
+        bool probePaused
+    ) internal view override returns (uint256 amountOut) {
+        if (poolModule == address(this)) {
+            // Check if pool is paused, if requested
+            if (probePaused) {
+                // We issue a static call in case the pool does not conform to IPausable interface.
+                (bool success, bytes memory returnData) = pool.staticcall(
+                    abi.encodeWithSelector(IPausable.paused.selector)
+                );
+                if (success && abi.decode(returnData, (bool))) {
+                    // Pool is paused, return zero
+                    return 0;
+                }
+            }
+            // Pool conforms to IDefaultPool interface.
+            try
+                IDefaultPool(pool).calculateSwap({
+                    tokenIndexFrom: tokenIndexes[pool][_nodes[nodeIndexFrom].token],
+                    tokenIndexTo: tokenIndexes[pool][_nodes[nodeIndexTo].token],
+                    dx: amountIn
+                })
+            returns (uint256 amountOut_) {
+                amountOut = amountOut_;
+            } catch {
+                // Return zero if the pool getter reverts for any reason
+                amountOut = 0;
+            }
+        } else {
+            // Ask Pool Module to calculate the quote
+            address tokenFrom = _nodes[nodeIndexFrom].token;
+            address tokenTo = _nodes[nodeIndexTo].token;
+            // Here we pass both token address and its index to the pool module, so it doesn't need to store
+            // index<>token mapping. This allows Pool Module to be implemented in a stateless way, as some
+            // pools require token index for interactions, while others require token address.
+            try
+                IPoolModule(poolModule).getPoolQuote(
+                    pool,
+                    IndexedToken({index: tokenIndexes[pool][tokenFrom], token: tokenFrom}),
+                    IndexedToken({index: tokenIndexes[pool][tokenTo], token: tokenTo}),
+                    amountIn,
+                    probePaused
+                )
+            returns (uint256 amountOut_) {
+                amountOut = amountOut_;
+            } catch {
+                // Return zero if the pool module getter reverts for any reason
+                amountOut = 0;
+            }
+        }
+    }
+
+    /// @dev Returns the tokens in the pool at the given address.
+    function _getPoolTokens(
+        address poolModule,
+        address pool,
+        uint256 tokensAmount
```

Comment:
There's no explicit check that `tokensAmount` actually equals the number of coins in the pool. While `onlyOwner` makes this somewhat ok, could become an issue when adding pools if the wrong param is given, particularly given the pool module implementations are rather flexible

Reply:
That's a very good point. I feel there are two possible ways to approach this:
1. Have the owner specify number of tokens and the pool, but verify this during pool addition (requires every pool module to check this in `getPoolTokens()`).
2. Don't specify the number of tokens in the pool, and make this a pool module responsibility to fetch ALL the pool tokens in `getPoolTokens()`.

(1) saves a bit of gas, but I feel that (2) is easier to enforce throughout the codebase, so leaning towards (2). @mikeyrf which one do you prefer?

Reply:
Adjusted in f63fc7908ca39755369818fa90b962f9d5ceb9c4, with a breaking change for Pool Module interface.